### ISSUE TITLE
Zone aware topologySpreadConstraints not working

### ISSUE THREAD

**Issue by ejsmith** [elastic/cloud-on-k8s#6066]
## Bug Report
**What did you do?**
Followed instructions for making the ES cluster zone aware.

**What did you expect to see?**
Pods scheduled evenly across all zones.

**What did you see instead? Under which circumstances?**
3 total zones, 3 node master cluster, 1 zone with 2 pods each, 1 zone with 0 pods

**Environment**
AWS EKS

* ECK version:
2.4.0

* Kubernetes information:
Cloud: EKS
Version: v1.23.7

* Resource definition:
```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  annotations:
    eck.k8s.elastic.co/downward-node-labels: "topology.kubernetes.io/zone"
  name: ats-perf
spec:
  version: 7.17.6
  image: slideroom/elasticsearch:7.17.6 # https://github.com/slideroom/infrastructure/blob/master/docker/elasticsearch
  http:
    service:
      metadata:
        annotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
          service.beta.kubernetes.io/aws-load-balancer-internal: "true"
      spec:
        type: LoadBalancer
    tls:
      selfSignedCertificate:
        # subjectAltNames:
        # - dns: elastic.ats.int
        # - dns: localhost
        disabled: true
  nodeSets:
  - name: master-v1
    count: 3
    config:
      node.roles: [ "master" ]
      node.attr.zone: ${ZONE}
      cluster.routing.allocation.awareness.attributes: k8s_node_name,zone
      xpack.monitoring.collection.enabled: true
      reindex.remote.whitelist: "elastic.ats.int:9200"
      action.destructive_requires_name: true
      # if not setting max_map_count in an init container, then use this setting
      #node.store.allow_mmap: false
    podTemplate:
      spec:
        containers:
        - name: elasticsearch
          env:
          - name: ES_JAVA_OPTS
            value: -Xms1g -Xmx1g
          - name: ZONE
            valueFrom:
              fieldRef:
                fieldPath: metadata.annotations['topology.kubernetes.io/zone']
          resources:
            limits:
              memory: 2Gi
              cpu: 1
        nodeSelector:
          role: elastic-master
        topologySpreadConstraints:
          - maxSkew: 1
            minDomains: 3
            topologyKey: topology.kubernetes.io/zone
            whenUnsatisfiable: DoNotSchedule
            labelSelector:
              matchLabels:
                elasticsearch.k8s.elastic.co/cluster-name: ats-perf
                elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-master-v1
        serviceAccountName: ats-perf-elastic
        # related to "node.store.allow_mmap: false" setting above
        initContainers:
        - name: sysctl
          securityContext:
            privileged: true
          command: ['sh', '-c', 'sysctl -w vm.max_map_count=262144']
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 4Gi
        storageClassName: io2-max-encrypted
  - name: data-v1
    count: 5
    config:
      node.roles: [ "data", "ingest", "transform" ]
      node.attr.zone: ${ZONE}
      cluster.routing.allocation.awareness.attributes: k8s_node_name,zone
      xpack.monitoring.collection.enabled: true
      reindex.remote.whitelist: "elastic.ats.int:9200"
      action.destructive_requires_name: true
      # if not setting max_map_count in an init container, then use this setting
      #node.store.allow_mmap: false
    podTemplate:
      spec:
        containers:
        - name: elasticsearch
          env:
          - name: ES_JAVA_OPTS
            value: -Xms13g -Xmx13g
          - name: ZONE
            valueFrom:
              fieldRef:
                fieldPath: metadata.annotations['topology.kubernetes.io/zone']
          resources:
            requests:
              memory: 26Gi
              cpu: 5700m
            limits:
              memory: 26Gi
              cpu: 5700m
        nodeSelector:
          role: elastic-data
        topologySpreadConstraints:
          - maxSkew: 1
            minDomains: 4
            topologyKey: topology.kubernetes.io/zone
            whenUnsatisfiable: DoNotSchedule
            labelSelector:
              matchLabels:
                elasticsearch.k8s.elastic.co/cluster-name: ats-perf
                elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-data-v1
        serviceAccountName: ats-perf-elastic
        # related to "node.store.allow_mmap: false" setting above
        initContainers:
        - name: sysctl
          securityContext:
            privileged: true
          command: ['sh', '-c', 'sysctl -w vm.max_map_count=262144']
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 100Gi
        storageClassName: io2-10-encrypted

---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
metadata:
  name: ats-perf
spec:
  version: 7.17.6
  count: 1
  elasticsearchRef:
    name: ats-perf
  config:
    xpack.monitoring.enabled: true
  podTemplate:
    spec:
      containers:
      - name: kibana
        env:
          - name: NODE_OPTIONS
            value: "--max-old-space-size=2048"
        resources:
          requests:
            memory: 1Gi
            cpu: 0.5
          limits:
            memory: 2.5Gi
            cpu: 2
  http:
    service:
      metadata:
        annotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
          service.beta.kubernetes.io/aws-load-balancer-internal: "true"
      spec:
        type: LoadBalancer
    tls:
      selfSignedCertificate:
        disabled: true
```

The node groups are set to autoscale. They were all scaled to 1 node each before and once 2 pods got scheduled on 1 of the zones the autoscaler scaled the other node back since it didn't have anything on it.

![image](https://user-images.githubusercontent.com/282584/193947176-35d50797-701e-4324-8100-3c806281f326.png)

**Comment by ejsmith** [comment]
Was wondering if anyone got a chance to look at this? Am I doing something wrong?

**Comment by naemono** [comment]
I am working on replicating this issue in EKS, and will update as soon as I have more information.

**Comment by ejsmith** [comment]
Thanks! Any help would be greatly appreciated.

**Comment by naemono** [comment]
@ejsmith Unfortunately not quite yet.  I haven't forgotten about this, and will be getting back to it as soon as I can.

**Comment by barkbay** [comment]
```
        nodeSelector:
          role: elastic-master
```

Since you specified a `nodeSelector` it would be helpful to check that this label is correctly set on the K8S nodes. If not it could explain why the `topologySpreadConstraints` is not behaving as expected.

**Comment by ejsmith** [comment]
> ```
> 
>         nodeSelector:
> 
>           role: elastic-master
> 
> ```
> 
> 
> 
> Since you specified a `nodeSelector` it would be helpful to check that this label is correctly set on the K8S nodes. If not it could explain why the `topologySpreadConstraints` is not behaving as expected.

Yes, I did verify that the role attribute was being set.

**Comment by barkbay** [comment]
> Yes, I did verify that the role attribute was being set.

Would it be possible to share the list nodes with the labels? It would help to replicate the issue.

Note that Pod scheduling is the done by Kubernetes itself, not by ECK. If the `labelSelector` fields are correct (which seem to be the case) and the `StatefulSets`  do contain the expected `topologySpreadConstraints` , then my first impression is that this behaviour is related to K8S itself and not ECK. In other words, for now, I would expect the same behavior outside of ECK.

nit: I think `minDomains` is only available by default since `1.25`, did you enable the related feature gate?

**Comment by ejsmith** [comment]
Here is the node lists using a label filter to show that there are nodes with those labels:

![image](https://user-images.githubusercontent.com/282584/198329190-aa2aec61-3e36-4568-b9ba-e09abcdf8f9a.png)

Here is the stateful set with some sections removed to make it shorter:

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  labels:
    common.k8s.elastic.co/template-hash: "2975646590"
    common.k8s.elastic.co/type: elasticsearch
    elasticsearch.k8s.elastic.co/cluster-name: ats-perf
    elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-master-v1
  name: ats-perf-es-master-v1
  namespace: default
spec:
  replicas: 3
  selector:
    matchLabels:
      common.k8s.elastic.co/type: elasticsearch
      elasticsearch.k8s.elastic.co/cluster-name: ats-perf
      elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-master-v1
  serviceName: ats-perf-es-master-v1
  template:
    metadata:
      annotations:
        co.elastic.logs/module: elasticsearch
        elasticsearch.k8s.elastic.co/config-hash: "201090172"
      creationTimestamp: null
      labels:
        common.k8s.elastic.co/type: elasticsearch
        elasticsearch.k8s.elastic.co/cluster-name: ats-perf
        elasticsearch.k8s.elastic.co/http-scheme: http
        elasticsearch.k8s.elastic.co/node-data: "false"
        elasticsearch.k8s.elastic.co/node-data_cold: "false"
        elasticsearch.k8s.elastic.co/node-data_content: "false"
        elasticsearch.k8s.elastic.co/node-data_frozen: "false"
        elasticsearch.k8s.elastic.co/node-data_hot: "false"
        elasticsearch.k8s.elastic.co/node-data_warm: "false"
        elasticsearch.k8s.elastic.co/node-ingest: "false"
        elasticsearch.k8s.elastic.co/node-master: "true"
        elasticsearch.k8s.elastic.co/node-ml: "false"
        elasticsearch.k8s.elastic.co/node-remote_cluster_client: "false"
        elasticsearch.k8s.elastic.co/node-transform: "false"
        elasticsearch.k8s.elastic.co/node-voting_only: "false"
        elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-master-v1
        elasticsearch.k8s.elastic.co/version: 7.17.6
    spec:
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchLabels:
                  elasticsearch.k8s.elastic.co/cluster-name: ats-perf
              topologyKey: kubernetes.io/hostname
            weight: 100
      automountServiceAccountToken: false
      dnsPolicy: ClusterFirst
      nodeSelector:
        role: elastic-master
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: ats-perf-elastic
      serviceAccountName: ats-perf-elastic
      terminationGracePeriodSeconds: 180
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            elasticsearch.k8s.elastic.co/cluster-name: ats-perf
            elasticsearch.k8s.elastic.co/statefulset-name: ats-perf-es-master-v1
        maxSkew: 1
        topologyKey: topology.kubernetes.io/zone
        whenUnsatisfiable: DoNotSchedule
  updateStrategy:
    type: OnDelete
  volumeClaimTemplates:
  - apiVersion: v1
    kind: PersistentVolumeClaim
    metadata:
      creationTimestamp: null
      name: elasticsearch-data
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 4Gi
      storageClassName: io2-max-encrypted
      volumeMode: Filesystem
    status:
      phase: Pending
status:
  availableReplicas: 3
  collisionCount: 0
  currentReplicas: 3
  currentRevision: ats-perf-es-master-v1-d697b8ffd
  observedGeneration: 3
  readyReplicas: 3
  replicas: 3
  updateRevision: ats-perf-es-master-v1-d697b8ffd
  updatedReplicas: 3
```

I didn't realize `minDomains` required 1.25, so that explains why that part didn't work.

**Comment by naemono** [comment]
Sorry it took me so long to get back to this issue, but I can't replicate this issue.

Nodes: 2 nodegroups with role labels set and with 3x nodes spread across AZs
```
❯ kubectl get nodes -o wide -L "topology.kubernetes.io/zone" -L role
NAME                             STATUS   ROLES    AGE   VERSION               INTERNAL-IP      EXTERNAL-IP      OS-IMAGE         KERNEL-VERSION                 CONTAINER-RUNTIME   ZONE         ROLE
ip-192-168-15-242.ec2.internal   Ready    <none>   21m   v1.23.9-eks-ba74326   192.168.15.242   ---    Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1a   elastic-master
ip-192-168-35-187.ec2.internal   Ready    <none>   8d    v1.23.9-eks-ba74326   192.168.35.187   ---   Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1b   elastic-master
ip-192-168-38-99.ec2.internal    Ready    <none>   11m   v1.23.9-eks-ba74326   192.168.38.99    ---     Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1b   elastic-data
ip-192-168-4-78.ec2.internal     Ready    <none>   11m   v1.23.9-eks-ba74326   192.168.4.78     ---     Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1a   elastic-data
ip-192-168-66-78.ec2.internal    Ready    <none>   11m   v1.23.9-eks-ba74326   192.168.66.78    ---      Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1c   elastic-data
ip-192-168-82-86.ec2.internal    Ready    <none>   21m   v1.23.9-eks-ba74326   192.168.82.86    ---     Amazon Linux 2   5.4.209-116.367.amzn2.x86_64   docker://20.10.17   us-east-1c   elastic-master
```

I apply the (slightly modified) elasticsearch.yaml and I get pods spread evenly across AZs

pods:
```
❯ kubectl get pods -n default -o json | jq '.items[]|{ name: .metadata.name, zone: .metadata.annotations["topology.kubernetes.io/zone"] }'
{
  "name": "ats-perf-es-data-v1-0",
  "zone": "us-east-1a"
}
{
  "name": "ats-perf-es-data-v1-1",
  "zone": "us-east-1c"
}
{
  "name": "ats-perf-es-data-v1-2",
  "zone": "us-east-1b"
}
{
  "name": "ats-perf-es-master-v1-0",
  "zone": "us-east-1c"
}
{
  "name": "ats-perf-es-master-v1-1",
  "zone": "us-east-1a"
}
{
  "name": "ats-perf-es-master-v1-2",
  "zone": "us-east-1b"
}
{
  "name": "ats-perf-kb-5bb67d4c64-8qbqp",
  "zone": null
}
```

Could we get the events in this namespace, which should show the `Scheduled`, or `FailedScheduling` events that correspond to why each pod was put where it was?

```
kc get event -n default | grep Sched
```

I do want to mirror exactly what @barkbay pointed out `Pod scheduling is the done by Kubernetes itself, not by ECK`.  From what I see, ECK is placing the proper `topologySpreadConstraints`, `nodeSelector`, and `affinity` fields on the appropriate `Statefulsets`.

**Comment by barkbay** [comment]
I'm closing because of inactivity and also because the operator is not involved in Pod scheduling operations. I would first check if, apart from any ECK considerations, the K8S Scheduler schedules the Pods as expected.

That being said feel free to reopen if you have any additional information to reproduce and if you think that there is indeed a bug in the operator.

**Comment by cwiggs** [comment]
I ran into a similar (maybe the same issue) issue recently.  As far as I can tell the issue comes down to the persistent volume.  Here is what I've seen:

1. Create a node group in just AZ-1, with 2 nodes.
2. Deploy a ECK nodeSet (count = 2) with the proper zone topology set.
3. ECK will deploy 2 nodes to the same AZ since it doesn't have an option for a 2nd AZ.
4. Create a 2nd node pool, with 1 node in AZ-2
5. At this point restarting the statefulSet, killing any of the pods in the statefulSet, etc will just have a node start up in AZ-1.  If you try and force a pod onto the worker node in AZ-2 it will fail saying that the PV has an affinity for AZ-1.  The only way I have found to "fix" this issue is to delete the PVC and PV, then restart the pod, or statefulSet.  Another option is to bring up a new nodeSet to replace the old one.

At some point I would expect the ECK operator to know that it needs to update the statefulSet or PV with a new affinity for the 2nd AZ.  The Operator could then create a new PV in the 2nd AZ, migrate the data off of the old PV in the first AZ, then kill the old PV/PVC/pod that is in the first AZ.

I'm running my ECK deployment in OCI's OKE so it is different but I'm willing to help troubleshoot this issue, post my manifests, etc if needed, but I think the numbered list above should enable you to re-produce the issue.

Thanks
